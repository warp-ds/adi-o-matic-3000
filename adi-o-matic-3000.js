import yaml from 'yaml'
const collected = {}

function adiOMatic3000(csv) {

  const lines = csv.split('\n')
  for (const line of lines) {
    const [name, value] = line.split(',')
    const segments = name.split(/[-|.]+/)

    segments.reduce((acc, currentValue, index) => {
      const lastItem = index + 1 === segments.length;

      // set value as string or object containg the value and previous values depending on last item type
      if(lastItem) {
        if(typeof acc[currentValue] === 'object') {
          return acc[currentValue] = { ...acc[currentValue], "_": value }
        }

        return acc[currentValue] = value
      }

      // rewrite currentValue as an object if its a string
      if(typeof acc[currentValue] === 'string') {
        return acc[currentValue] = { "_": acc[currentValue] }
      }

      return acc[currentValue] = acc[currentValue] || {};
    }, collected);
  }

  return yaml.stringify(collected)
}

export { adiOMatic3000 }
