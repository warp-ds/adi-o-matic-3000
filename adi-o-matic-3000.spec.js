import fs from 'node:fs'
import { expect, test } from "vitest";
import { adiOMatic3000 } from "./adi-o-matic-3000";

const csv = fs.readFileSync('./mock-data.csv', 'utf-8')

test("adi-o-matic 3000 generates correct yaml", () => {
  const yaml = adiOMatic3000(csv)

  expect(yaml).toMatchSnapshot()
});
