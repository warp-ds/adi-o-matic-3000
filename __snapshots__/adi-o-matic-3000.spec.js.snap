// Vitest Snapshot v1

exports[`adi-o-matic 3000 generates correct yaml 1`] = `
"color:
  token:
    of:
      greatness: blue-500
      doom: red-500
    _: foo
  button:
    primary:
      wombat:
        text: aqua-600
"
`;
